8349146 [REDO] Implement a better allocator for downcalls

minborg opened the pull request:
This PR is based on the work of @mernst-github and aims to implement a thread-local 'stack' allocator, which works like a dynamically sized arena, but with reset functionality to reset the allocated size back to a certain level. The underlying memory could stay around between calls, which could improve performance.

minborg commented:
/contributor add @mernst-github

minborg commented:
> Can we just cleanly revert commit [7764742](https://github.com/openjdk/jdk/commit/77647421c5b13738f6fa145bf30cbb9dd741e164)?
> This new version defines redundant sealed interface + record - feels unnecessary to me.
> Meanwhile, the BufferStackBench is valueable, but its addition should be kept in a separate commit to make reviewing easier.

The idea was to use records for trusted components and less boilerplate. The idea with the interface was to hide the record accessors a bit.

liach commented:
`trust_final_non_static_fields` in `ciField` already trusts `jdk.internal.foreign`, so I don't think the record performance argument stands here. But making Frame and BufferStack classes final is a good move.

liach reviewed (approved):
Looks to be in parity with the original patch.

JornVernee commented:
We can not just re-use the original test due to: https://bugs.openjdk.org/browse/JDK-8350455

See my comment here: https://github.com/openjdk/jdk/pull/23142#issuecomment-2673194510

JornVernee commented:
> maybe we should problem list or exclude macOS testing?

Yes, I think we should. The issue is quite intermittent, and last time it only started showing up in CI as well. I think we should move the `stress()` test to a separate file./jtreg test, and then problem list that test on Mac, with a link to JDK-8350455

JornVernee commented:
It looks like you've made some other changes since the first patch. It would be useful if you could summarize those changes.

minborg commented:
So, the recent changes relate mostly to improving test coverage and some fixes that were made as a result:
 * Reused tests from another PR with similar ambitions
 * Added convenience methods for being able to use a MemoryLayout directly
 * Added thread check in `Frame::allocate" preventing inadvertent sharing across threads
 * Broke out a test and marked it in a ProblemList (macosx malloc problem)
 * Adding overrides to the usual `Object` methods for `BufferStack`

minborg commented:
Updated benchmarks:

```
Benchmark                            (ELEM_SIZE)  Mode  Cnt   Score   Error  Units
BufferStackBench.OfVirtual.buffer              8  avgt   15  12.653 ± 0.189  ns/op
BufferStackBench.OfVirtual.buffer             16  avgt   15  12.635 ± 0.228  ns/op
BufferStackBench.OfVirtual.buffer             32  avgt   15  12.562 ± 0.062  ns/op
BufferStackBench.OfVirtual.confined            8  avgt   15  22.649 ± 0.262  ns/op
BufferStackBench.OfVirtual.confined           16  avgt   15  23.507 ± 0.146  ns/op
BufferStackBench.OfVirtual.confined           32  avgt   15  25.426 ± 0.512  ns/op
BufferStackBench.buffer                        8  avgt   15   4.693 ± 0.037  ns/op
BufferStackBench.buffer                       16  avgt   15   4.715 ± 0.032  ns/op
BufferStackBench.buffer                       32  avgt   15   4.670 ± 0.038  ns/op
BufferStackBench.confined                      8  avgt   15  22.548 ± 0.022  ns/op
BufferStackBench.confined                     16  avgt   15  23.494 ± 0.092  ns/op
BufferStackBench.confined                     32  avgt   15  25.285 ± 0.076  ns/op
```

and

```
Benchmark                              Mode  Cnt   Score   Error  Units
CallOverheadByValue.OfVirtual.byPtr    avgt   15   4.095 ± 0.035  ns/op
CallOverheadByValue.OfVirtual.byValue  avgt   15  16.666 ± 0.208  ns/op
CallOverheadByValue.byPtr              avgt   15   4.132 ± 0.146  ns/op
CallOverheadByValue.byValue            avgt   15  11.835 ± 0.139  ns/op
```

Pases `tier1` testing on various platforms and JVM configurations

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/BufferStack.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Override
+    public int hashCode() {
+        return System.identityHashCode(this);
```
What's the value in these overloads? Isn't this implementation the same we'd get if there were no overrides?

mcimadamore replied:
Ugh - it's a record, I see. Which brings up the question -- should it be a record? Should it advertize its `size and `tl` components (they seem rather implementation specific?).

mcimadamore replied:
Was this done so that the final field could be trusted? (I believe all final fields are implicitly trusted in jdk/internal/foreign):

```
      holder->is_in_package("jdk/internal/foreign/layout") || holder->is_in_package("jdk/internal/foreign") ||
```

minborg replied:
Yes. We should revert to a normal class.

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/BufferStack.java:
```diff
@@ -0,0 +1,188 @@
+                if (locked) {
+                    lock.unlock();
+                }
+                Reference.reachabilityFence(arena);
```
I'm not sure this is enough to keep the automatic arena alive. If the client lets the Frame arena go out of scope w/o calling close, then `arena` will become unreachable, but some segments created by the Frame arena might still be reachable. To be more correct, I think `Frame` should add a "close action" to its confined arena which keeps the outer automatic arena alive. This can be done, for instance, by passing a close action to the `reinterpret` call:

```
frame = new SlicingAllocator(frameSegment.reinterpret(confinedArena, () -> Reference.reachabilityFence(arena)));
```

The close action is installed in the `MemorySession` object of `confinedArena` -- which is then attached to all segments returned by `Frame` -- thus keeping the automatic arena alive.

mcimadamore replied:
(if you agree with this analysis, perhaps adding an extra stress test, or tweaking one of the existing stress tests to check this could also be useful)

JornVernee replied:
I've thought about it a bit, and I think adding a cleanup action to the reinterpret call is indeed the only fool proof way to add a link back to the parent segment. MemorySegment is what gives a user access to the memory (almost like an access token), so the memory needs to be kept alive through the MS. MS has a reference to its scope, which can then reference the original scope through a cleanup action.

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/BufferStack.java:
```diff
@@ -0,0 +1,188 @@
+            @SuppressWarnings("restricted")
+            public MemorySegment allocate(long byteSize, long byteAlignment) {
+                // Make sure we are on the right thread
+                if (((MemorySessionImpl) scope()).ownerThread() != Thread.currentThread()) {
```
This can also be rewritten as:
`MemorySession.toMemorySession(confinedArena).checkValidState()` -- which is what we use elsewhere

mcimadamore commented on test/jdk/java/foreign/TestBufferStack.java:
```diff
@@ -0,0 +1,281 @@
+                var segment = arena.allocate(JAVA_INT);
+                assertEquals(firstAddress, segment.address());
+                var segmentTwo = arena.allocate(JAVA_INT);
+                assertNotEquals(firstAddress, segmentTwo.address());
```
Should this check be stronger -- e.g. that the address of the second segment should be the address of the first + the size of JAVA_INT?

mcimadamore commented on test/jdk/java/foreign/TestBufferStack.java:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    @Test
+    void allocationCaptureStateLayout() {
```
This test is not really testing much re. capture state layout. It feels like the only added thing w.r.t. `allocationSameAsPoolSize` is that it also checks that additional allocation will fail with IOOBE -- maybe these two tests should be merged?

mcimadamore commented on test/jdk/java/foreign/TestBufferStack.java:
```diff
@@ -0,0 +1,281 @@
+
+    @Test
+    void closeConfinement() {
+        var pool = BufferStack.of(POOL_SIZE);
```
Nice confinement tests!

mcimadamore commented on test/jdk/java/foreign/TestBufferStack.java:
```diff
@@ -0,0 +1,281 @@
+import static java.lang.foreign.ValueLayout.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+final class TestBufferStack extends NativeTestHelper {
```
This also extends `NativeTestHelper` -- does it need to?

mcimadamore commented on test/jdk/java/foreign/TestBufferStackStress2.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.concurrent.ForkJoinWorkerThread;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+final class TestBufferStackStress2 extends NativeTestHelper {
```
This also extends `NativeTestHelper` -- does it need to?

mcimadamore commented on test/jdk/java/foreign/TestBufferStackStress.java:
```diff
@@ -0,0 +1,72 @@
+ * @test
+ * @modules java.base/jdk.internal.foreign
+ * @build NativeTestHelper TestBufferStackStress
+ * @run junit/othervm --enable-native-access=ALL-UNNAMED TestBufferStackStress
```
is `--enable-native-access` needed here (and also in the other stress test) ?

mcimadamore commented on test/micro/org/openjdk/bench/java/lang/foreign/CallOverheadByValue.java:
```diff
@@ -0,0 +1,100 @@
+    @Benchmark
+    public void byValue() throws Throwable {
+        // point = unit();
+        MemorySegment unused = (MemorySegment) MH_UNIT_BY_VALUE.invokeExact(
```
This benchmark is a bit misleading, because the allocator object will add some noise in the mix. I suggest to have some allocator object ready in a field, and just pass that (avoiding the lambda). That would also be more similar to idiomatic FFM code (which assumes some allocator is already available at the callsite).

minborg commented on test/jdk/java/foreign/TestBufferStackStress2.java:
Good catch! I will fix that and also add a check that the VT is completed as expected to avoid such issues in the future.

wenshao commented on src/java.base/share/classes/jdk/internal/foreign/BufferStack.java:
```diff
@@ -0,0 +1,188 @@
+        }
+
+        static PerThread of(long size) {
+            final Arena arena = Arena.ofAuto();
+            final SlicingAllocator stack = new SlicingAllocator(arena.allocate(size));
+            return new PerThread(new ReentrantLock(), stack, arena);
```
```suggestion
            final Arena arena = Arena.ofAuto();
            return new PerThread(new ReentrantLock(), 
                                 new SlicingAllocator(arena.allocate(size)), arena);
        }
```
The local variable stack is only used once, is it okay not to use it?

minborg commented:
Updated benchmarks:

```
Benchmark                            (ELEM_SIZE)  Mode  Cnt   Score   Error  Units
BufferStackBench.OfVirtual.buffer              8  avgt   15  12.695 ± 0.205  ns/op
BufferStackBench.OfVirtual.buffer             16  avgt   15  12.622 ± 0.078  ns/op
BufferStackBench.OfVirtual.buffer             32  avgt   15  12.523 ± 0.022  ns/op
BufferStackBench.OfVirtual.confined            8  avgt   15  22.902 ± 0.407  ns/op
BufferStackBench.OfVirtual.confined           16  avgt   15  23.858 ± 0.652  ns/op
BufferStackBench.OfVirtual.confined           32  avgt   15  25.544 ± 0.458  ns/op
BufferStackBench.buffer                        8  avgt   15   4.923 ± 0.029  ns/op
BufferStackBench.buffer                       16  avgt   15   4.971 ± 0.095  ns/op
BufferStackBench.buffer                       32  avgt   15   4.980 ± 0.105  ns/op
BufferStackBench.confined                      8  avgt   15  22.713 ± 0.289  ns/op
BufferStackBench.confined                     16  avgt   15  23.576 ± 0.348  ns/op
BufferStackBench.confined                     32  avgt   15  25.272 ± 0.530  ns/op
```

and

```
vBenchmark                              Mode  Cnt   Score   Error  Units
CallOverheadByValue.OfVirtual.byPtr    avgt   15   4.159 ± 0.097  ns/op
CallOverheadByValue.OfVirtual.byValue  avgt   15  16.535 ± 0.433  ns/op
CallOverheadByValue.byPtr              avgt   15   4.119 ± 0.031  ns/op
CallOverheadByValue.byValue            avgt   15  11.404 ± 0.240  ns/op
```

wenshao commented on src/java.base/share/classes/jdk/internal/foreign/BufferStack.java:
```diff
@@ -0,0 +1,206 @@
+                }
+            }
+        }
+    }
```
```suggestion
    private record PerThread(ReentrantLock lock, Arena arena, SlicingAllocator stack) {

        @ForceInline
        @SuppressWarnings("restricted")
        public Arena pushFrame(long size, long byteAlignment) {
            boolean needsLock = Thread.currentThread().isVirtual() && !lock.isHeldByCurrentThread();
            if (needsLock && !lock.tryLock()) {
                // Rare: another virtual thread on the same carrier competed for acquisition.
                return Arena.ofConfined();
            }
            if (!stack.canAllocate(size, byteAlignment)) {
                if (needsLock) lock.unlock();
                return Arena.ofConfined();
            }

            Arena confinedArena = Arena.ofConfined();
            long parentOffset = stack.currentOffset();
            MemorySegment frameSegment = stack.allocate(size, byteAlignment);
            long topOfStack = stack.currentOffset();
            // The cleanup action will keep the original automatic `arena` (from which
            // the reusable segment is first allocated) alive even if this Frame
            // becomes unreachable but there are reachable segments still alive.
            SegmentAllocator frame = new SlicingAllocator(frameSegment.reinterpret(confinedArena, new Frame.CleanupAction(arena)));

            return new Frame(this, needsLock, parentOffset, topOfStack, confinedArena, frame);
        }

        static PerThread of(long byteSize, long byteAlignment) {
            final Arena arena = Arena.ofAuto();
            return new PerThread(new ReentrantLock(),
                    arena,
                    new SlicingAllocator(arena.allocate(byteSize, byteAlignment)));
        }
    }

    private record Frame(PerThread thread, boolean locked, long parentOffset, long topOfStack, Arena confinedArena, SegmentAllocator frame) implements Arena {
        record CleanupAction(Arena arena) implements Consumer<MemorySegment> {
            @Override
            public void accept(MemorySegment memorySegment) {
                Reference.reachabilityFence(arena);
            }
        }

        @ForceInline
        private void assertOrder() {
            if (topOfStack != thread.stack.currentOffset())
                throw new IllegalStateException("Out of order access: frame not top-of-stack");
        }

        @ForceInline
        @Override
        @SuppressWarnings("restricted")
        public MemorySegment allocate(long byteSize, long byteAlignment) {
            // Make sure we are on the right thread and not closed
            MemorySessionImpl.toMemorySession(confinedArena).checkValidState();
            return frame.allocate(byteSize, byteAlignment);
        }

        @ForceInline
        @Override
        public MemorySegment.Scope scope() {
            return confinedArena.scope();
        }

        @ForceInline
        @Override
        public void close() {
            assertOrder();
            // the Arena::close method is called "early" as it checks thread
            // confinement and crucially before any mutation of the internal
            // state takes place.
            confinedArena.close();
            thread.stack.resetTo(parentOffset);
            if (locked) {
                thread.lock.unlock();
            }
        }
    }
```
I find it a bit strange to nest a class inside a record. What do you think if I change it to two records?

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/BufferStack.java:
```diff
@@ -102,27 +119,41 @@ public Arena pushFrame(long size, long byteAlignment) {
+            return new PerThread(new ReentrantLock(),
+                    arena,
+                    new SlicingAllocator(arena.allocate(byteSize, byteAlignment)),
+                    new CleanupAction(arena));
```
any reason why you didn't use a lambda here?

mcimadamore replied:
Also, not a big fan of records here -- it seems that many implementation details such as cleanup action, lock and slicing allocator are "leaked out" to the caller, that is now responsible to set things up correctly. I think a `PerThread` constructor with arena, size, alignment would be make the code less coupled and more readable.

mcimadamore replied:
And, if you do that, you then don't need the `of` factory -- clients can just use `new`

wenshao replied:
> any reason why you didn't use a lambda here?

I also think that CleanupAction should be changed to lambda

minborg replied:
Using an anonymous class for the cleanup action had very adverse effects on performance. I didn't want to use a lambda for startup performance reasons.

wenshao replied:
If using lambda affects performance, how about using anonymous classes?
```java
            return new PerThread(new ReentrantLock(),
                    arena,
                    new SlicingAllocator(arena.allocate(byteSize, byteAlignment)),
                    new Consumer<MemorySegment>() {
                        @Override
                        public void accept(MemorySegment memorySegment) {
                            Reference.reachabilityFence(arena);
                        }});
```

liach replied:
Anonymous classes also captures outer variables and break cleaner/GC. The only safe measures are local enums or records (anonymous classes cannot be enum or records)

mcimadamore commented:
> Updated benchmarks:
> 
> ```
> Benchmark                            (ELEM_SIZE)  Mode  Cnt   Score   Error  Units
> BufferStackBench.OfVirtual.buffer              8  avgt   15  12.695 ± 0.205  ns/op
> BufferStackBench.OfVirtual.buffer             16  avgt   15  12.622 ± 0.078  ns/op
> BufferStackBench.OfVirtual.buffer             32  avgt   15  12.523 ± 0.022  ns/op
> BufferStackBench.OfVirtual.confined            8  avgt   15  22.902 ± 0.407  ns/op
> BufferStackBench.OfVirtual.confined           16  avgt   15  23.858 ± 0.652  ns/op
> BufferStackBench.OfVirtual.confined           32  avgt   15  25.544 ± 0.458  ns/op
> BufferStackBench.buffer                        8  avgt   15   4.923 ± 0.029  ns/op
> BufferStackBench.buffer                       16  avgt   15   4.971 ± 0.095  ns/op
> BufferStackBench.buffer                       32  avgt   15   4.980 ± 0.105  ns/op
> BufferStackBench.confined                      8  avgt   15  22.713 ± 0.289  ns/op
> BufferStackBench.confined                     16  avgt   15  23.576 ± 0.348  ns/op
> BufferStackBench.confined                     32  avgt   15  25.272 ± 0.530  ns/op
> ```
> 
> and
> 
> ```
> Benchmark                              Mode  Cnt   Score   Error  Units
> CallOverheadByValue.OfVirtual.byPtr    avgt   15   4.159 ± 0.097  ns/op
> CallOverheadByValue.OfVirtual.byValue  avgt   15  16.535 ± 0.433  ns/op
> CallOverheadByValue.byPtr              avgt   15   4.119 ± 0.031  ns/op
> CallOverheadByValue.byValue            avgt   15  11.404 ± 0.240  ns/op
> ```

What was the result before?

mcimadamore reviewed (approved):
Looks good - I left some stylistic comments on the use of records

minborg commented:
The performance before this PR can be seen in the "confined" benchmarks above. In those benchmarks, a regular `Arena.ofConfined` is created upon every invocation.

minborg commented:
/integrate

I will integrate this PR now @wenshao . Can you summarize your proposed changes below as It was a bit unclear to me what you meant. I can create a separate PR with those changes later.

wenshao commented:
> Can you summarize your proposed changes below as It was a bit unclear to me what you meant. I can create a separate PR with those changes later.

My suggestion is to use two records (both PerThread and Frame) to replace the original inner class Frame. This would also eliminate one use of @ForceInline. This might just be my personal coding style preference – feel free to adopt it if you like. Just a suggestion!

mcimadamore commented:
> The performance before this PR can be seen in the "confined" benchmarks above. In those benchmarks, a regular `Arena.ofConfined` is created upon every invocation.

Yes - but I was referring to the performance of the native call -- not that of the allocation per se (e.g. `CallOverheadByValue`). If I'm not mistaken the numbers of that benchmark all refer to new code (well, except when we pass by pointer, in which case there's no allocation -- both before and after this PR).
